x/incentives/keeper/pot.go

```diff
@@ -0,0 +1,308 @@
+			if amt.IsPositive() {
+				distrCoins = distrCoins.Add(sdk.NewCoin(coin.Denom, amt))
+			}
+		}
+		distrCoins = distrCoins.Sort()
+		if distrCoins.Empty() {
+			continue
+		}
+		if err := k.bk.SendCoinsFromModuleToAccount(ctx, types.ModuleName, lock.Owner, distrCoins); err != nil {
+			return nil, err
+		}
+		totalDistrCoins = totalDistrCoins.Add(distrCoins...)
+	}
+
+	// increase filled epochs after distribution
+	pot.FilledEpochs += 1
+	pot.DistributedCoins = pot.DistributedCoins.Add(totalDistrCoins...)
+	k.setPot(ctx, &pot)
+
+	return totalDistrCoins, nil
+}
+
+// GetModuleToDistributeCoins returns sum of to distribute coins for all of the module
+func (k Keeper) GetModuleToDistributeCoins(ctx sdk.Context) sdk.Coins {
+	activePotsDistr := k.getToDistributeCoinsFromIterator(ctx, k.ActivePotsIterator(ctx))
+	upcomingPotsDistr := k.getToDistributeCoinsFromIterator(ctx, k.UpcomingPotsIteratorAfterTime(ctx, ctx.BlockTime()))
+	return activePotsDistr.Add(upcomingPotsDistr...)
+}
+
+// GetModuleDistributedCoins returns sum of distributed coins so far
+func (k Keeper) GetModuleDistributedCoins(ctx sdk.Context) sdk.Coins {
+	activePotsDistr := k.getDistributedCoinsFromIterator(ctx, k.ActivePotsIterator(ctx))
+	finishedPotsDistr := k.getDistributedCoinsFromIterator(ctx, k.FinishedPotsIterator(ctx))
+	return activePotsDistr.Add(finishedPotsDistr...)
+}
+
+// GetPotByID Returns pot from pot ID
+func (k Keeper) GetPotByID(ctx sdk.Context, potID uint64) (*types.Pot, error) {
+	pot := types.Pot{}
+	store := ctx.KVStore(k.storeKey)
+	potKey := potStoreKey(potID)
+	if !store.Has(potKey) {
+		return nil, fmt.Errorf("pot with ID %d does not exist", potID)
+	}
+	bz := store.Get(potKey)
+	k.cdc.MustUnmarshalJSON(bz, &pot)
+	return &pot, nil
+}
+
+// GetPots returns pots both upcoming and active
+func (k Keeper) GetPots(ctx sdk.Context) []types.Pot {
+	return append(k.GetActivePots(ctx), k.GetUpcomingPots(ctx)...)
+}
+
+// GetActivePots returns active pots
+func (k Keeper) GetActivePots(ctx sdk.Context) []types.Pot {
+	return k.getPotsFromIterator(ctx, k.ActivePotsIterator(ctx))
+}
+
+// GetUpcomingPots returns scheduled pots
+func (k Keeper) GetUpcomingPots(ctx sdk.Context) []types.Pot {
+	return k.getPotsFromIterator(ctx, k.UpcomingPotsIterator(ctx))
+}
+
+// GetFinishedPots returns finished pots
+func (k Keeper) GetFinishedPots(ctx sdk.Context) []types.Pot {
+	return k.getPotsFromIterator(ctx, k.FinishedPotsIterator(ctx))
+}
+
+// GetRewardsEst returns rewards estimation at a future specific time
+func (k Keeper) GetRewardsEst(ctx sdk.Context, addr sdk.AccAddress, locks []lockuptypes.PeriodLock, pots []types.Pot, endEpoch int64) sdk.Coins {
+	// initialize pots to active and upcomings if not set
+	if len(pots) == 0 {
+		pots = k.GetPots(ctx)
+	}
+
+	// estimate rewards
+	estimatedRewards := sdk.Coins{}
+	params := k.GetParams(ctx)
+	currentEpoch, _ := k.GetCurrentEpochInfo(ctx)
+
+	cacheCtx, _ := ctx.CacheContext()
```

Comment:
nicer to document why to use cached context

Reply:
got it